todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/TasksMvpController.java

```diff
@@ -0,0 +1,214 @@
+     */
+    public TaskDetailFragment createDetailViewForTablet(String taskId) {
+        // Create the View
+        TaskDetailFragment taskDetailFragment = TaskDetailFragment.newInstance(taskId);
+
+        // Create the Presenter
+        TaskDetailPresenter taskDetailPresenter = createTaskDetailPresenter(taskId,
+                taskDetailFragment);
+
+        // Wire presenters
+        mTasksTabletPresenter.setTaskDetailPresenter(taskDetailPresenter);
+        taskDetailFragment.setPresenter(mTasksTabletPresenter);
+        return taskDetailFragment;
+    }
+
+    private void createTabletElements() {
+        TasksFragment tasksFragment = findOrCreateTasksFragment();
+
+        // TasksFragment is retained so let's reuse its presenter.
+        if (tasksFragment.getPresenter() != null) {
+            TasksTabletPresenter retainedPresenter =
+                    (TasksTabletPresenter) tasksFragment.getPresenter();
+            mTasksPresenter = retainedPresenter.getTasksPresenter();
+        } else {
+            mTasksPresenter = createListPresenter(tasksFragment);
+        }
+
+        mTasksTabletPresenter = new TasksTabletPresenter(
+                Injection.provideTasksRepository(mFragmentActivity),
+                new TasksTabletNavigator(mFragmentActivity, this),
+                mTasksPresenter);
+
+        tasksFragment.setPresenter(mTasksTabletPresenter);
+
+        // TaskDetailFragment is retained, so let's reuse its presenter.
+        TaskDetailFragment taskDetailFragment = getDetailFragment();
+        if (taskDetailFragment != null && taskDetailFragment.isAdded()) {
+
+            if (taskDetailFragment.getPresenter() instanceof TasksTabletPresenter) {
+                TasksTabletPresenter retainedPresenter =
+                        (TasksTabletPresenter) taskDetailFragment.getPresenter();
+
+                TaskDetailPresenter retainedTaskDetailPresenter =
+                        retainedPresenter.getTaskDetailPresenter();
+
+                mTasksTabletPresenter.setTaskDetailPresenter(retainedTaskDetailPresenter);
+
+                // Replace retained presenter with new one.
+                taskDetailFragment.setPresenter(mTasksTabletPresenter);
+            }
+        }
+    }
+
+    @NonNull
+    private TaskDetailPresenter createTaskDetailPresenter(
+            String taskId, TaskDetailFragment taskDetailFragment) {
+        return new TaskDetailPresenter(
+                taskId,
+                Injection.provideTasksRepository(mFragmentActivity.getApplicationContext()),
+                taskDetailFragment);
+    }
+
+    private TasksPresenter createListPresenter(TasksFragment tasksFragment) {
+        mTasksPresenter = new TasksPresenter(
+                Injection.provideTasksRepository(mFragmentActivity.getApplicationContext()),
+                tasksFragment);
+
+        return mTasksPresenter;
+    }
+
+    @NonNull
+    private TasksFragment findOrCreateTasksFragment() {
```

Comment:
The name of this method is `findOrCreateTaskFragment`, understanding that will implement those operations in the method body (find or create a fragment). 

Actually, is doing one more action not specified in the method signature, which is 'add the fragment to an activity', so I would suggest changing the method's name to  **findOrAddTaskFragment**